File: src/main/java/com/gitlab/daring/image/component/BaseCanvasFrame.java
package com.gitlab.daring.image.component;

import org.bytedeco.javacpp.opencv_core.Mat;
import org.bytedeco.javacv.CanvasFrame;
import org.bytedeco.javacv.Frame;
import org.bytedeco.javacv.OpenCVFrameConverter.ToMat;

import java.awt.*;

import static com.gitlab.daring.image.swing.SwingUtils.runInEdt;
import static com.gitlab.daring.image.util.GeometryUtils.scaleToMax;

public class BaseCanvasFrame extends CanvasFrame {

    final ToMat converter = new ToMat();
    final Dimension maxSize = new Dimension(1024, 768);

    public BaseCanvasFrame(String title) {
        super(title, 1);
        setVisible(false);
    }

    public void showMat(Mat m) {
        Frame cm = converter.convert(m);
        runInEdt(() -> showImage(cm));
    }

    @Override
    public void showImage(Image image) {
        if (!isVisible()) needInitialResize = true;
        super.showImage(image);
        if (!isVisible()) {
            resize();
            setVisible(true);
        }
    }

    public void resize() {
        Dimension d = scaleToMax(getCanvasSize(), maxSize);
        setCanvasSize(d.width, d.height);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/swing/BaseCanvasFrame.kt
package com.gitlab.daring.image.swing

import com.gitlab.daring.image.swing.SwingUtils.runInEdt
import com.gitlab.daring.image.util.GeometryUtils.scaleToMax
import org.bytedeco.javacpp.opencv_core.Mat
import org.bytedeco.javacv.CanvasFrame
import org.bytedeco.javacv.OpenCVFrameConverter.ToMat
import java.awt.Dimension
import java.awt.Image

class BaseCanvasFrame(title: String) : CanvasFrame(title, 1.0) {

    val matConverter = ToMat()
    val maxSize = Dimension(1024, 768)

    init {
        isVisible = false
    }

    fun showMat(m: Mat) {
        val cm = matConverter.convert(m)
        runInEdt { showImage(cm) }
    }

    override fun showImage(image: Image?) {
        if (!isVisible) needInitialResize = true
        super.showImage(image)
        if (!isVisible) {
            resize()
            isVisible = true
        }
    }

    fun resize() {
        val d = scaleToMax(canvasSize, maxSize)
        setCanvasSize(d.width, d.height)
    }

}
